testing.md

```diff
@@ -7,22 +7,218 @@ recipes cover how to write automated tests and see their code
+
+The behavior of suite/test runner can be customized by the `*on-failure*` variable, which controls what to do when check failure happens. It can be set to one of the following values:
+- `:debug` to drop to debugger.
+- `:backtrace` to print a backtrace.
+- `NIl` (default) to simply continue.
+
+### ASDF integration
+
+So it would be nice to provide a one-line trigger to test our `my-system` system. Recall that we said it is better to provide a root suite? Here is the reason:
+
+~~~lisp
+(defsystem my-system
+  ;; Parts omitted.
+  :in-order-to ((test-op (test-op :my-system/test))))
+
+(defsystem mitogrator/test
+  ;; Parts omitted.
+  :perform (test-op (op c)
+                    (symbol-call :fiveam :run!
+                                 (find-symbol* :my-system :my-system/test))))
+~~~
+
+The last line tells ASDF to load symbol `:my-system` from `my-system/test` package and call `fiveam:run!`. It fact, it is equivalent to `(run! 'my-system)` as mentioned above.
+
+### Testing report customization
+
+It is possible to generate our own testing report. The macro `run!` is nothing more than a composition of `explain!` and `run`. 
+
+Instead of generating a testing report like its cousin `run!`, the function `run` runs suite or test passed in and returns a list of `test-result` instance, usually instances of `test-failure` or `test-passed` sub-classes.
+
+A class `text-explainer` is defined as a basic class for testing report generator. A generic function `explain` is defined to take a `text-plainer` instance and a `test-result` instance (returned by `run`) and generate testing report. The following 2 code snippets are equivalent:
+
+~~~lisp
+(run! 'read-file-as-string-non-existing-file)
+
+(explain (make-instance '5am::detailed-text-explainer)
+         (run 'read-file-as-string-non-existing-file))
+~~~
+
+By creating a new sub-class of `text-explainer` and a method `explain` for it, it is possible to define a new test reporting system.
+
+The following code just provides a proof-of-concept implementation. You may need to read the source code of `5am::detailed-text-explainer` to fully understand it.
+
+~~~lisp
+(defclass my-explainer (5am::text-explainer)
+  ())
+
+(defmethod 5am:explain ((explainer my-explainer) results &optional (stream *standard-output*) recursive-deps)
+  (loop for result in results
+        do (case (type-of result)
+             ('5am::test-passed
+              (format stream "~%Test ~a passed" (5am::name (5am::test-case result))))
+             ('5am::test-failure
+              (format stream "~%Test ~a failed" (5am::name (5am::test-case result)))))))
+
+(explain (make-instace 'my-explainer)
+         (run 'read-file-as-string-non-existing-file))
+; Test READ-FILE-AS-STRING-NON-EXISTING-FILE failed
+; Test READ-FILE-AS-STRING-NON-EXISTING-FILE passed => NIL
+~~~
+
+## Mocking with Mockingbird
+
+TODO
 
 ## Testing with Prove
 
+> Warning: Prove is obsolete. This section is kept but it is not encouraged to use it.
+
```

Comment:
It's OK to remove it, isn't it?

Reply:
> Yipee, thanks for tackling this issue. I think you can enable GitHub pages on your fork, so we could read this chapter for real and see how it looks? (like on https://vindarel.github.io/cl-cookbook/)

Oh! Okay..I will try that :-)

Reply:
I am not sure about this. I think Prove and Rove shares many similar (or even same?) APIs. Maybe it should be somehow transformed to Rove intro?

I have not used Rove for years and do not know its status now.

What do you think?

Reply:
No strong opinion. But Prove clearly is deprecated, its repository is archived. If you don't feel like updating the section for Rove in the same PR, we can delete the Prove section, and dig it up later© for Rove if needed.

Reply:
Hi all :)

I was looking at the page, and for complete beginners, it is more informative and gives a better outlook if we state that it's superseded by Rove, by the same author.

Something like:

> Warning: Prove is obsolete and superseded by [Rove](https://github.com/fukamachi/rove), by the same author. Both have similar APIs. This section is kept as an overview of Prove until a rewrite is done for Rove.

Reply:
This is actually the initial intention of mine, i.e. to keep it as a reference for Rove.

But now I am concerned that their APIs are somehow different. If so, this content might be a bit misleading...

@vindarel You make the call.

Reply:
I think it's well written and works for Prove, which isn't that different from Rove.

If you leave it in I might update it for Rove in the next weeks ;)

Reply:
If another voice says it's better for beginners to keep the Prove, so let's keep it, thanks for the feedback.

But then to properly close the issue we will need to 1) update it to Rove obviously and 2) compare the two frameworks and say what advantages there are to use Rove. The one I saw, using it from the command line, doesn't work as advertised.